sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventProcessorClient.java

```diff
@@ -283,13 +290,36 @@ public synchronized void start() {
      * <!-- end com.azure.messaging.eventhubs.eventprocessorclient.startstop -->
      */
     public synchronized void stop() {
+        try {
+            stop(DEFAULT_STOP_TIMEOUT);
+        } catch (RuntimeException e) {
+            logger.info("Error while stopping the event processor", e);
+        }
+    }
+
+    /**
+     * Stops processing events for all partitions owned by this event processor. All {@link PartitionProcessor} will be
+     * shutdown and any open resources will be closed.
+     *
+     * <p>
+     * Subsequent calls to stop will be ignored if the event processor is not running or is being stopped.
+     * </p>
+     *
+     * @param timeout The maximum amount of time to wait for the processor to stop processing.
+     *
+     * @throws RuntimeException if the event processor encounters timout or another error while stopping.
+     */
+    public synchronized void stop(Duration timeout) {
         if (!isRunning.compareAndSet(true, false)) {
             logger.info("Event processor has already stopped");
             return;
         }
         runner.get().cancel(true);
+
         scheduler.get().shutdown();
-        stopProcessing();
+
+        Mono.when(Mono.fromCallable(() -> scheduler.get().awaitTermination(timeout.toMillis(), TimeUnit.MILLISECONDS)), stopProcessing())
+            .block(timeout);
```

Comment:
I was wondering about the use of the same `timeout` value as the timeout for inner publisher and outer publisher composing that inner? should the inner timeout is shorter than the outer, just thinking loud.

Reply:
these tasks would normally run in parallel, so the worst that can happen if that outer timeout would fire before the inner one. It can't change anything for the `awaitTermination` task so `awaitTermination` will just run its course and throw internally too.

It seems the worst that could happen is more exceptions being thrown and possibly logged, but it's the end of the processor life and is not a problem. So I'd prefer to keep things simple. 

I'm happy to learn if there is any problem with my logic!

Reply:
Thank you for the brain storming, I agree, the outer timeout won’t impact the scheduler resource cleanup, so this logic is simple and serves the purpose.